### DISCUSSION TITLE
Reconnect RDP session on double click

### DISCUSSION THREAD

**Discussion by MWinhard** [1Remote/1Remote#354] [Ideas]
**Is your feature request related to a problem? Please describe.**
Situation: PC went to sleep. I switched it on again. RDP connection is disconnected and can not be automatically restored.
Current behavior: I have to right click the connection title, then left click "Reconnect". That's too much clicking.

**Describe the solution you'd like**
Reconnect on double click on the RDP connection title.

**Additional context**
My RDP connections are running in a window (NOT full screen).

![image](https://user-images.githubusercontent.com/111646687/213365703-c7043c8e-bc13-43e3-a8b6-5b63304c03c9.png)

**Comment by majkinetor** [comment]
Isn't reconnect button in the middle of the screen also, requiring 1 click ?


Be that as it may, it is probably good idea to have this _as an option_, to automatically reconnect after some time. Probably a global option, that can be also set on connection (which uses global option if it doesn't have its own).

**Comment by VShawn** [comment]
I dont think we need a double click reconnect, reconnect is not a high-frequency operation for most people, it's ok to make a complex tigger for it. May be we should focus on how to make it auto reconnect when PC awake form sleep. (I didn't it won't reconnect after sleep since all my pc are set to never sleep.)

Also what's the purpose for `automatically reconnect after some time` and how many people need this? imagine we are getting fun with remote PC when suddenly a reconnection was tiggered. 

Noted we can not get mouse and keyboard event in normal way when all this device was redirect to remote, which means we dont know user is AKF or enjoying the remote desk when PRM get no mouse event for a long time.

**Comment by majkinetor** [comment]
> Also what's the purpose for automatically reconnect after some time and how many people need this?

Purpose would be bad network connections that do not resolve quickly. VPN can choke also, happened to me all the time last year.

> Noted we can not get mouse and keyboard event in normal way when all this device was redirect to remote, which means we dont know user is AKF or enjoying the remote desk when PRM get no mouse event for a long time.

Indeed, which means you can not blindly do so but must know reason for previous termination. If user is on remote desk you wont get connection problems.

**Comment by MWinhard** [comment]
> Isn't reconnect button in the middle of the screen also, requiring 1 click ?

No it isn't. The Windows logon screen is shown.

**Comment by MWinhard** [comment]
An by the way: Such a long discussion for one line of code? Especially as it's good practice in Windows for many many years, that a double click triggers the default action.

![image](https://user-images.githubusercontent.com/111646687/213917761-8dd3c9b3-0217-4bac-bc93-bf69e584d2eb.png)

In this case the default action is Reconnect.

![image](https://user-images.githubusercontent.com/111646687/213917893-afcf2f28-ef98-418a-a664-70ce378cd816.png)

**Comment by majkinetor** [comment]
Yeah, such a long dicussion for any line of code. 

> No it isn't. The Windows logon screen is shown.

I don't recall I ever got windows logon screen. Only black window with reconnect button in the middle.

**Comment by MWinhard** [comment]
> I don't recall I ever got windows logon screen. Only black window with reconnect button in the middle.

You don't have to believe me. I'll post a screenshot when it happens next time. ;-)

**Comment by majkinetor** [comment]
I beilieve you, I just say I didn't see that :) 

@VShawn is this a bug then? Do you know something about difference in behavior?

**Comment by MWinhard** [comment]
Here's the promised screenshot. It happens usually once or twice a day.

![image](https://user-images.githubusercontent.com/111646687/214002370-7d8b73f8-a2c9-4c43-a5d2-ea6f53a8ee18.png)

To reconnect on doubleclick at the connection name button in the title bar would be a really convenient feature. Pleeeease... :-)

**Comment by majkinetor** [comment]
This is how it should look like:

![image](https://user-images.githubusercontent.com/85767/214003959-a634ddf9-0957-4d01-89f9-0fb6ee6a190a.png)

@VShawn, do you know why the difference? Looks like some RDP setting.

BTW, @MWinhard this is one click less then what you propose.

**Comment by MWinhard** [comment]
@majkinetor You're right, this page serves the same purpose. But I never saw it before.

BTW, the logon screen was diplayed when I was using mstsc.exe to a Win81 VM, too. So it's just RDP behavior. Nothing special with PRemoteM in this case.

Perhaps the different behavior on my side is triggered by this situation:
- company office is using a dyndns address
- I have to use a special port to connect to the Win11 VM in the company office
- my pc goes into hibernation mode
- company office gets a new ip address (usually every night, sometimes during the day, too)
- my pc is switched on (i.e. wakes up from hibernation)
- PRemoteM shows the logon screen of the Win11 VM

**Comment by majkinetor** [comment]
Can't you turn off hibernation to avoid getting new dyndns ?

I think its probably some RDP config on that particular server (or your companys group policy). It could probably be changed. 

But I understand your irritattion.... Is it possible to use launcher to recconect ? This should be as fast as with button? If that was last connection used, ALT+M+Enter would reconnect and its probably as fast as you can get. I can't try things because I don't know currently how to create such scenario to test.

**Comment by majkinetor** [comment]
@MWinhard, check this out: https://woshub.com/remote-desktop-session-time-limit
Specifically, _Automatically Log off Disconnected and Idle Remote Desktop User Sessions_

**Comment by VShawn** [comment]
I think it was the auto log off feature of the server. `reconnect` and `dismiss` shown only when the it was disconnected while in auto log off case the connection was still there so the login page was shown.

BTW I'm on a vacation now and I won't working on PRM before I back home. You can take your time to make discussion on this, I will check tickets on the phone now and then：）

**Comment by majkinetor** [comment]
I think it will be easy to do this by accident. Not something one would expect when doubleclicking. And if one is on SSH, it will lose shell session.

**Comment by MWinhard** [comment]
> I think it will be easy to do this by accident. Not something one would expect when doubleclicking.

Nothing bad happens when you Reconnect to an RDP session. So what?

> if one is on SSH, it will lose shell session.

That means for SSH sessions we have to ignore the double click.

Regarding your other suggestions:
- I won't turn off hibernation on my pc. It saves about 50 euro per month. Energy cost is quite high in my country these days.
- Using the launcher doesn't log me in again. I suppose it detects that the RDP session is already connected. And doesn't/can't detect that Windows is showing the logon screen.
- I'm a mouse user. Even if the launcher would work, it were even less convenient than right click and Reconnect. I activate the RDP window using the mouse. Now I have to move both hands to the keyboard, press Alt+M (or whatever hotkey), then the ENTER key. Then move my right hand back to the mouse. You see what I mean.
- Session timeout wouldn't solve my problem. There's often unfinished work open during the night. I want to finish it next morning.
- There is nothing at all configured with the RDP connection. Not in the Win11 VM and not on my Win10 PC. Everything Windows default behavior.

**Comment by VShawn** [comment]
> I think it will be easy to do this by accident. Not something one would expect when doubleclicking.

I agree with Mr. majkinetor, I don't want to waste any second dure to an accident double clicking. You see waste time is `something bad happens` to us.


===

I was just trying to monitor the login window, but as soon as it shows up, the 'reconnect' button pops up too.

However, when I read the document again trying to find out if we can detect when a login window is shown. Unfortunately it turn out that the wrapper of MSTSC offering by Microsoft can not do that :(

**Comment by MWinhard** [comment]
@VShawn What does this mean? You will not implement the requested double click feature?

**Comment by majkinetor** [comment]
It can't be implemented the way you want. Its not a good UX. 

It could eventually be implemented as 

1. an config option that would be by default disabled ( [ ] Connecton title doubleclick triggers reconnecting ). I still dont like it, totally arbitrary feature creep.
2. another button in the titlebar, also as an option, disabled by default ( [o] Show reconnect button in the connection title bar )
![image](https://user-images.githubusercontent.com/85767/215768865-81736fed-d656-45ca-b3f7-db210f47599d.png)

I think second option makes sense, since right clicking shows _close_ and _reconnect_ as options. Since you can close with middle click, you technically do not need x button (so it could be an option too).

So I think that we could have "Connection button options"

3. [ ] Edit button?
1. [ ] Reconnect button
2. [ ] Close button

and we can keep the context menu having all of the options always.


This would allow us to have some other buttons in the future evetnually, if the need arise (_edit_ for example, could be very useful for adding Markdown Notes as requested on another ticket).  This would let one have anything one wants, even no buttons at all.

**Comment by VShawn** [comment]
@MWinhard Yes I don't plan to follow your suggestion to fix this ticket as it goes against intuition and may cause misclicks. My original idea was to detect the login window but as mentioned, it doesn't work. My alternative solution is to add a button in the header, just as solution 2 by @majkinetor.

**Comment by MWinhard** [comment]
> 1. an config option that would be by default disabled ( [ ] Connecton title doubleclick triggers reconnecting ). I still dont like it, totally arbitrary feature creep.
> 2. another button in the titlebar, also as an option, disabled by default ( [o] Show reconnect button in the connection title bar )
>    ![image](https://user-images.githubusercontent.com/85767/215768865-81736fed-d656-45ca-b3f7-db210f47599d.png)

You can imagine I vote for option 1. ;-)

If it has to be option 2, how about the reconnect button on the left side of the connection title? Microsoft Office does this for years. As the buttons are quite small, keeping them separate minimizes the risk of accidently clicking the wrong button.

![image](https://user-images.githubusercontent.com/111646687/215780468-61002140-397d-4f19-8956-8807f326305d.png)

![image](https://user-images.githubusercontent.com/111646687/215780540-cad2fa5b-9db2-4257-a58a-a7a3422a8b79.png)

**Comment by majkinetor** [comment]
> You can imagine I vote for option 1. ;-)

No I can't. Its one click more then with the button :)

>  As the buttons are quite small, keeping them separate minimizes the risk of accidently clicking the wrong button

You would be able to disable all other buttons. The title of the Office tools are much much wider, it allows for such shineningas.

**Comment by VShawn** [comment]
> No I can't. Its one click more then with the button :)

I would like to add that double-clicking the title in Chrome browser closes the page. If changed to another function, it may cause confusion among Chrome users.

> how about the reconnect button on the left side of the connection title? 

I don't like having functional buttons in different places, I find it unattractive, sorry this is a matter of my personal taste :)
